tests/phpunit/tests/media.php

```diff
@@ -3698,6 +3693,140 @@ public function test_wp_omit_loading_attr_threshold() {
+
+	/**
+	 * Tests that wp_get_loading_attr_default() returns the expected loading attribute value before loop but after get_header in main query but header was not called.
+	 *
+	 * @ticket 58211
+	 *
+	 * @covers ::wp_get_loading_attr_default
+	 *
+	 * @dataProvider data_wp_get_loading_attr_default_before_and_no_loop
+	 *
+	 * @param string $context Context for the element for which the `loading` attribute value is requested.
+	 */
+	public function test_wp_get_loading_attr_default_before_loop_in_main_query_but_header_not_called( $context ) {
+		global $wp_query, $wp_the_query;
+
+		$this->get_new_wp_query_and_reset_variables();
+
+		// Set current query as main query.
+		$wp_the_query = $wp_query;
+
+		// Lazy if header not called.
+		$this->assertSame( 'lazy', wp_get_loading_attr_default( $context ) );
+	}
+
+	/**
+	 * Tests that wp_get_loading_attr_default() returns the expected loading attribute value before loop but after get_header for main query.
+	 *
+	 * @ticket 58211
+	 *
+	 * @covers ::wp_get_loading_attr_default
+	 *
+	 * @dataProvider data_wp_get_loading_attr_default_before_and_no_loop
+	 *
+	 * @param string $context Context for the element for which the `loading` attribute value is requested.
+	 */
+	public function test_wp_get_loading_attr_default_before_loop_if_main_query( $context ) {
+		global $wp_query, $wp_the_query;
+
+		$this->get_new_wp_query_and_reset_variables();
+
+		// Set current query as main query.
+		$wp_the_query = $wp_query;
+
+		do_action( 'get_header' );
+		$this->assertFalse( wp_get_loading_attr_default( $context ) );
+	}
+
+	/**
+	 * Tests that wp_get_loading_attr_default() returns the expected loading attribute value after get_header and after loop.
+	 *
+	 * @ticket 58211
+	 *
+	 * @covers ::wp_get_loading_attr_default
+	 *
+	 * @dataProvider data_wp_get_loading_attr_default_before_and_no_loop
+	 *
+	 * @param string $context Context for the element for which the `loading` attribute value is requested.
+	 */
+	public function test_wp_get_loading_attr_default_after_loop( $context ) {
+		global $wp_query, $wp_the_query;
+
+		$this->get_new_wp_query_and_reset_variables();
+
+		// Set current query as main query.
+		$wp_the_query = $wp_query;
+
+		do_action( 'get_header' );
+
+		while ( have_posts() ) {
+			the_post();
+		}
+		$this->assertSame( 'lazy', wp_get_loading_attr_default( 'wp_get_attachment_image' ) );
+	}
+
+	/**
+	 * Tests that wp_get_loading_attr_default() returns the expected loading attribute if no loop.
+	 *
+	 * @ticket 58211
+	 *
+	 * @covers ::wp_get_loading_attr_default
+	 *
+	 * @dataProvider data_wp_get_loading_attr_default_before_and_no_loop
+	 *
+	 * @param string $context Context for the element for which the `loading` attribute value is requested.
+	 */
+	public function test_wp_get_loading_attr_default_no_loop( $context ) {
+		global $wp_query, $wp_the_query;
+
+		$this->get_new_wp_query_and_reset_variables();
+
+		// Set current query as main query.
+		$wp_the_query = $wp_query;
```

Comment:
We don't set wp_query as main query in all the tests that uses new query.  e.g see above `test_wp_get_loading_attr_default_before_loop_if_not_main_query`

Reply:
Another option is to have one more function for `main query` that consumes this helper and add this extra line. And we can call the new function over here.  @spacedmonkey would like your opinion here.

Reply:
@kt-12 add a parameter to `get_new_wp_query_and_reset_variables`? 

```
protected function get_new_wp_query_and_reset_variables( $setup_main_query = true ){
....
if( $setup_main_query ) {
       // Set current query as main query.
       $wp_the_query = $wp_query;
}
```

In places where you don't to setup main query, do this. 
```
$this->get_new_wp_query_and_reset_variables( false );
```

Or in that one test, don't don't use that util method...

Reply:
Bundling all of this in a single test method doesn't really make sense to me. Getting a query is completely separate from resetting global state, and also completely separate from setting a query as the main query.

Yes, we shouldn't repeat code, but we also shouldn't violate the single responsibility principle. Adding a boolean parameter to change what a method does violates that principle too. Fewer lines of code is not the main goal, the goal is to have fewer lines of low-level code that increases the chance for error - calling a few methods isn't really an issue.

The original code here was already using helper methods `reset_content_media_count()` and `reset_omit_loading_attr_filter()`. If we want to also abstract out the `new WP_Query` part, we could introduce a method like `get_new_wp_query_for_published_post()`. And then we can also introduce another method `set_main_query( $query )` to avoid direct access to the globals in all the tests.

Then the tests could be something like this:
```php
$query = $this->get_new_wp_query_for_published_post();
$this->set_main_query( $query );
$this->reset_content_media_count();
$this->reset_omit_loading_attr_filter();
```

That way, every method has just one purpose, and having 3 or 4 of those lines repeated across tests is perfectly fine since there is pretty much zero room left for error.

Reply:
@spacedmonkey Can you please share your thoughts here on my feedback above? I don't think putting all of this in one method is the right approach.

Reply:
@felixarntz I agree that bundling unrelated function would be a bad practice. I have gone ahead and created utils the way you proposed.